src/Features/Core/Portable/Snippets/AbstractSnippetProvider.cs

```diff
@@ -0,0 +1,119 @@
+            var caretTarget = annotatedRootForCursor.GetAnnotatedNodes(_cursorAnnotation).SingleOrDefault();
+            return new Snippet(
+                snippetType: GetSnippetDisplayName(),
+                textChange: change,
+                cursorPosition: GetTargetCaretPosition(caretTarget),
+                renameLocations: await GetRenameLocationsAsync(reformattedDocument, span, cancellationToken).ConfigureAwait(false));
+        }
+
+        private static async Task<Document> GetDocumentWithSnippetAsync(Document document, TextSpan span, TextChange snippet, int tokenSpanStart, int tokenSpanEnd, CancellationToken cancellationToken)
+        {
+            var originalText = await document.GetTextAsync(cancellationToken).ConfigureAwait(false);
+
+            Document? snippetContainingDocument;
+
+            // Need to special case if the span is empty because otherwise it takes out preceding
+            // characters when we just want to remove any characters to invoke the completion.
+            // Example invoking when span is empty:
+            // public void Method()
+            // {
+            //     $$             <------ invoking by typing Ctrl-Space
+            // }
+            // Example invoking when span is not empty:
+            // public void Method()
+            // {
+            //     Wr$$           <----- invoking by typing out the completion 
+            // }
+            if (span.IsEmpty)
+            {
+                originalText = originalText.WithChanges(snippet);
+                snippetContainingDocument = document.WithText(originalText);
+            }
+            else
+            {
+                var textChange = new TextChange(TextSpan.FromBounds(tokenSpanStart, tokenSpanEnd), string.Empty);
+                originalText = originalText.WithChanges(textChange);
+                var document1 = document.WithText(originalText);
+                var textWithoutInvocation = await document1.GetTextAsync(cancellationToken).ConfigureAwait(false);
+                textWithoutInvocation = textWithoutInvocation.WithChanges(snippet);
+                snippetContainingDocument = document1.WithText(textWithoutInvocation);
+            }
+
+            return snippetContainingDocument;
+        }
+
+        private async Task<Document> ReformatDocumentAsync(Document document, TextSpan span, CancellationToken cancellationToken)
+        {
+            var annotatedSnippetRoot = await AnnotateRootForReformattingAsync(document, span, _reformatAnnotation, cancellationToken).ConfigureAwait(false);
+            document = document.WithSyntaxRoot(annotatedSnippetRoot);
+            return await Formatter.FormatAsync(document, _reformatAnnotation, cancellationToken: cancellationToken).ConfigureAwait(false);
+        }
+
+        public string GetSnippetText()
+        {
+            return GetSnippetDisplayName();
+        }
```

Comment:
Why the different names, and why bother implementing this here at all, instead of just making it abstract? At first glance seeing `ISnippetProvider.GetSnippetText()` I would expect it to contain `Console.WriteLine();`.. ie, the text that the snippet would insert.

Reply:
i agree. there's a lot of names/text/display stuff that is pretty confusing.